modelzoo/ESMM/train.py

```diff
@@ -0,0 +1,882 @@
+        print(f'\tWorkQueue: {args.workqueue}')
+    temp_optimizer = 'adam' if stock_tf else args.optimizer
+    print(f'Used optimizer: {temp_optimizer}')
+
+    # set fixed random seed
+    SEED = args.seed
+    tf.set_random_seed(SEED)
+
+    # set directory path for checkpoint_dir
+    keep_checkpoint_max = args.keep_checkpoint_max
+    model_dir = os.path.join(args.output_dir, 'model_ESMM_' + str(int(time.time())))
+    checkpoint_dir = args.checkpoint_dir if args.checkpoint_dir else model_dir
+    print(f'Saving model events to {checkpoint_dir}')
+
+    if keep_checkpoint_max:
+        print(f'Maximum number of saved checkpoints: {keep_checkpoint_max}')
+
+    # create data pipline of train & test dataset
+    train_dataset = build_model_input(train_file, batch_size, no_epochs, SEED, stock_tf, args.workqueue)
+    test_dataset = build_model_input(test_file, batch_size, 1, SEED, stock_tf, args.workqueue)
+
+    iterator = tf.data.Iterator.from_structure(train_dataset.output_types,
+                                               train_dataset.output_shapes)
+    next_element = iterator.get_next()
+
+    train_init_op = iterator.make_initializer(train_dataset)
+    test_init_op = iterator.make_initializer(test_dataset)
+
+    # create feature column
+    user_column, item_column, combo_column = build_feature_columns(stock_tf,
+                                                                   args.emb_fusion,
+                                                                   args.ev,
+                                                                   args.ev_elimination,
+                                                                   args.ev_filter,
+                                                                   args.adaptive_emb,
+                                                                   args.dynamic_ev)
+
+    # create variable partitioner for distributed training
+    num_ps_replicas = len(tf_config['ps_hosts']) if tf_config else 0
+    input_layer_partitioner = partitioned_variables.min_max_variable_partitioner(
+                                max_partitions=num_ps_replicas,
+                                min_slice_size=args.input_layer_partitioner <<
+                                    20) if args.input_layer_partitioner else None
+    dense_layer_partitioner = partitioned_variables.min_max_variable_partitioner(
+                                max_partitions=num_ps_replicas,
+                                min_slice_size=args.dense_layer_partitioner <<
+                                    10) if args.dense_layer_partitioner else None
+
+    # Session config
+    sess_config = tf.ConfigProto()
+    sess_config.inter_op_parallelism_threads = args.inter
+    sess_config.intra_op_parallelism_threads = args.intra
+
+    # Session hook
+    hooks = []
+
+    if args.smartstaged and not stock_tf:
+        '''SmartStage Feature'''
+        next_element = tf.staged(next_element, num_threads=8, capacity=40)
```

Comment:
set num_threads=4. It must be smaller than inter/intra and host's cores. Our benchmark machine is 8vCPU, 8cores.